docs/design/feature-attribute-semantics.md

```diff
@@ -0,0 +1,200 @@
+
+### Virtual methods
+
+- Overriding a _feature requirement_ method with a method outside of a _feature available_ scope is disallowed.
+- Overriding a method outside of a _feature available_ scope with a _feature requirement_ method is disallowed.
+
+### Member access
+
+Access to a _feature requirement_ method, constructor, field, property, or event outside of a _feature available_ scope is disallowed.
+
+## Feature checks
+
+Some feature attributes also come with corresponding feature checks that can be evaluated as constant at the time of trimming, with the guarded code removed when a feature is disabled. This effectivtely places the guarded code in a _feature available_ scope for the purposes of this analysis. However, the definition of such feature checks is left unspecified for now.
+
+## Trimming
+
+These semantics have been designed with trimming in mind. When a feature is disabled (by user configuration, or based on limitations of the target platform), trimming an app that will remove most or all of the feature-related code. Specifically, when a feature is disabled and an app has no trim warnings (including suppressed warnings):
+
+- Methods, fields, properties, and events which have a _feature requirement_ are guaranteed to be removed.
+
+- Methods which are in a _feature available_ scope but aren't entirely removed are guaranteed to have the method body replaced with a throwing instruction sequence.
+
+Thie latter can happen for methods in a type with _feature requirement_ (but that do not themselves have _feature requirement_) that are referenced outside of a _feature available_ scope. The reference to such a method may remain even though the type is never constructed. The callsite would produce a `NullReferenceException` and the method body is unreachable.
+
+## Alternatives
+
+One simplification would be to unify the concepts of _feature requirement_ with _feature available_, and treat both as similar to preprocessor symbols, where _any_ reference to a guarded type or member from an unguarded context is disallowed.
+
+The advantage of the specified model is that it allows some references without warning, giving some extra flexibility and making it easier to migrate existing code. The downside is that it might lead to preserving more code, whereas a simplified model could guarantee that all code related to a disabled feature is removed.
+
+Here is an example of a pattern which does not warn in the current model, but would warn with a simplified model. Assume that the code under `SomeFeatureIsSupported` is removed when the feature is unavailable.
+
+```csharp
+class FeatureConsumer {
+    static void Run() {
+        SomeFeatureProvider? some;
+        if (Features.SomeFeatureIsSupported)
```

Comment:
If the API to check for feature/capability availability is uniform, this could be completely removed, right? Then we would be able to use the simplified model and not have the extra warnings? Not that that's how it works right now, but it could be a possibility.

Reply:
Oh, I see, I missed that, thank you!